На пути к совершенству

kind-chigik opened the pull request:
---
:mortar_board: [На пути к совершенству](https://up.htmlacademy.ru/yii/1/user/669277/tasks/6)

kind-chigik committed:
Созданы абстрактный класс и его потомки

Для действий "Отменить", "Завершить", "Отказаться", "Откликнуться" создано по отдельному классу.

kind-chigik committed:
Добавлен метод для получения доступных действий, внесены корректировки

- Добавлен метод getAvailableActions, определяющий доступные действия в зависимости от статуса задачи и id пользователя.

- В массиве доступных действий ACTION теперь хранятся полные неймспейсы классов действий.

- В массиве статусов STATUS исправлено название ключа для статуса "Отменено" на более подходящее.

- Добавлено новое свойство $idCurrentUser для хранения текущего id пользователя

- В конструктор теперь передается массив, содержащий значения id пользователей.

kind-chigik committed:
Добавлен тестовый сценарий для метода getAvailableActions, внесены сопутствующие правки

- Добавлена новая переменная для хранения id текущего пользователя.

- Для переменных, содержащих id пользователей создан массив. Теперь именно его передаем в конструктор.

- Добавлен тестовый сценарий для проверки метода getAvailableActions. Проверяем правильность созданных объектов в зависимости от статуса и id пользователя.

Ovsyanka commented on src/Actions/CancelAction.php:
```diff
@@ -0,0 +1,27 @@
+
+    public static function checkRightUser($idCustomer, $idCurrentUser, $status)
+    {
+        if (($TaskStrategy->idCurrentUser === $TaskStrategy->idCustomer) && ($status === TaskStrategy::STATUS['new'])) {
```
А $TaskStrategy тут не определена. И еще переменные надо называть с маленькой буквы, а с большой - классы.

kind-chigik committed:
Исправлена ошибка с передачей данных в функции

Теперь в функции getAvailableActions и checkRightUser передается объект класса TaskStrategy.

Ovsyanka commented on src/TaskStrategy.php:
```diff
@@ -2,37 +2,42 @@
-
-    public function __construct($idCustomer, $idExecutor, $timeEnd, $statusActive)
+ 
+    public function __construct($idUser, $timeEnd, $statusActive)
```
А не лучше ли idUser тут передавать тремя разными аргументами? Вообще, название аргумента в таком случае вводит в заблуждение - думаешь, что там id юзера какой-то, а там массив определенной (неизвестной, так как нет документации к функции) структуры. Нужно смотреть исходный код функции, чтобы понять, что он должен из себя представлять.

Ovsyanka commented on src/TaskStrategy.php:
```diff
@@ -2,37 +2,42 @@
+use TaskForce\Actions\RefuseAction;
+use TaskForce\Actions\RespondAction;
+
 class TaskStrategy {
```
Было бы круто, если бы ты написала описание для класса - что он из себя представляет и какие задачи должен выполнять.

Ovsyanka replied:
Почему я об этом написал? Не совсем понятно, как ты видишь этот класс и, соответственно, на какие аспекты нужно обращать внимание при ревью. Например, я так понимаю, что этот класс должен определять возможные действия и изменения статусов для конкретной задачи в конкретном состоянии. В зависимости от этого у него должны быть разные функции с их аргументами и свойства.

Ovsyanka commented on src/TaskStrategy.php:
```diff
@@ -2,37 +2,42 @@
     public $idExecutor;
+    public $idCurrentUser;
     public $timeEnd;
     public $statusActive;
```
Это свойство не используется? Если оно не нужно - его стоит убрать. Вообще, по идее, оно должно бы использоваться в getAvailableActions - ведь доступные действия должны зависить от текущего статуса. И в таком случае его стоит назвать по-другому. statusActive переводится как "статус Активен". В то время как имеется ввиду, наверное, "активный(текущий) статус", то есть currentStatus, например. Или просто status.

Ovsyanka commented on src/TaskStrategy.php:
```diff
@@ -68,4 +73,38 @@ public function getNextStatus($action)
-}
\ No newline at end of file
+
+    public function getAvailableActions(TaskStrategy $obj, $status)
```
Смущает меня тут передача статуса через аргумент и экземпляра TaskStrategy тоже.
Исходя из моего текущего понимания класса - статус должен быть его свойством, соответственно, его не надо передавать. Но если я его понимаю не правильно - в другом случае статус передавать может быть нужно.
А вот с передачей экземпляра самого же TaskStrategy мне не понятно совсем. Что функция делает? Стоит написать к ней документацию, которая описывает что именно принимает/возвращает функция.

kind-chigik committed:
Изменены аргументы функции, добавлен новый класс действия

Из аргументов функции checkRightUser удален аргумент типа объект. Вместо него передаются id пользователей.

$status переименована в $currentStatus для наглядности.

Добавлен новый класс AcceptAction для действия "Принять"

kind-chigik committed:
Исправлен метод получения действий, внесены корректировки

- В константу ACTION добавлен новый объект действия "Принять".

- Изменена видимость для свойств класса. Теперь они приватные.

- Удалено лишнее свойство $idCurrentUser.

- Из аргументов конструктора удален массив c id пользователей. Вместо него теперь все id передаются отдельными переменными.

- Функция getAvailableActions теперь принимает только один аргумент - id текущего пользователя. Скорректированы аргументы для статичного метода checkRightUser. Добавлено условие для создания объекта действия "Принять".

kind-chigik committed:
Переписан код теста для проверки метода getAvailableActions

Ovsyanka commented on src/Actions/AbstractActions.php:
```diff
@@ -1,10 +1,9 @@
     abstract public function getNameAction();
     abstract public function getInsideName();
-    abstract public static function checkRightUser(TaskStrategy $obj, $status);
+    abstract public static function checkRightUser($idUser, $idCurrentUser, $status);
```
Под $idUser в дочерних классах подразумевается разное. Сейчас получается, что для вызова этой функции нужно знать ее внутреннее устройство - ид какого юзера в ней используется. То есть по сути интерфейс у них разный получается, так быть не должно.

Для определения того, можно ли выполнить действие нам нужно знать
1. Роль пользователя в рамкох данной задачи
2. Текущий статус

Первый пункт можно реализовать по-разному:
* Передавать идентификаторы трех пользователей - автор, исполнитель, текущий и внутри функции уже на основе них определять кем является текущий пользователь
* Сразу передавать роль пользователя: автор/исполнитель/никто.

Мы предполагали использовать ~~второй~~ первый вариант.

Ovsyanka replied:
Кстати, в ТЗ написано: "В классе каждого действия есть метод проверки прав. Этот метод принимает id исполнителя задания, id заказчика задания и id текущего пользователя. Метод возвращает true или false в зависимости от доступности выполнения этого действия. "

Ovsyanka commented on test.php:
```diff
@@ -1,16 +1,19 @@
-];
+$idCustomer = 128;
+$idExecutor = 127;
+$idCurrentUser = 128;
```
Переменная не используется, удалить.

Ovsyanka commented on test.php:
```diff
@@ -44,48 +47,104 @@
-    } else {
-        print 'Объект создан неверно';
-    }
+print ('Проверяем связку [Статус - Новое, Пользователь - Неавторизованный]' . '<br>');
```
Неавторизованный пользователь - это пользователь с неизвестным id (то есть отсутствующим). А тут, видимо, пользователь, который не является "участником" задачи. Можно его называть "Не участник", "Никто" (в контексте его роли в задаче), "Без роли".

Ovsyanka commented on test.php:
```diff
@@ -44,48 +47,104 @@
+            $result = true;
+        } else {
+            $result = false;
+        break;
```
Тут можно было бы просто сделать return.

Ovsyanka commented on test.php:
```diff
@@ -44,48 +47,104 @@
+ * @return bool true если искомое значение есть в массиве, в котором происходит поиск, иначе false и выход из цикла
+ */
+
+function areArrayEqual($currentActions, $сase) {
```
Документация! Крутяк =) Отмечу, что уже по описанию функции можно понять, что она не совсем подходит для использования в данных тестах. Если бы документации не было - мне было бы не понятно, например, в чем ты ошиблась - в постановке задачи или в ее реализации. Потому что я бы видел только реализацию.

1. Почему она не подходит. Что если у нас getAvailableActions вернет, скажем [RefuseAction], а в expected у нас [RefuseAction, CompleteAction]?

2. на мой взгляд, функция слишком привязана к вызывающему коду без особых оснований.

2.1 $case - зачем передавать сюда структуру из которой используется только одно значение (expected)? Лучше его сразу и передавать. Кроме того, это связано с документацией к функции. Сейчас она не корректная, получается и описывать то, что параметр должен являться массивом со свойством expected, которое в свою очередь является массивом, с которым сравнивается $currentActions... ради чего?

2.2. Имя первого аргумента $currentActions. А что поменяется в функционировании этой функции ,если там будет любой массив, не actions? Ничего. По крайней мере на данный момент. Вот если бы в ней было что-то такое, специфичное именно для массивов с действиями - тогда да, это было бы правильно.

3. Мне нравится имя функции. Оно подразумевает то, что будет возвращено булевое значение и это хорошо. Правда, корректней написать areArraysEqual. (arrays - множественное число, а то так получается "Является ли массив равнозначным" =))

Ovsyanka commented on test.php:
```diff
@@ -44,48 +47,104 @@
+ *
+ * @param array $currentActions Массив с проверяемыми значениями
+ * @param array $сase Массив, в котором должны содержаться искомые значения
+ * @return bool true если искомое значение есть в массиве, в котором происходит поиск, иначе false и выход из цикла
```
Реализацию тут описывать не надо: " и выход из цикла" - оно тут ни к чему, это внутренности функции.

Ovsyanka commented on test.php:
```diff
@@ -44,48 +47,104 @@
+        }
     }
+    return $result;
 }
```
Нет перехода строки в конце файла

Ovsyanka commented on test.php:
```diff
@@ -44,48 +47,104 @@
+    print 'Проверка для связки (Статус - Новое, Пользователь - Исполнитель) провалена' . '<br>';
+}
+
+print ('Проверяем связку [Статус - В работе, Пользователь - Заказчик]' . '<br>');
```
Много дублирования кода. В этом блоке изменяется совсем не много:
1. currentStatus
2. currentUser
3. описание проверки "(Статус - Новое, Пользователь - Заказчик)". Его, кстати, можно и автоматически сгенерить.
То есть по сути 2-3 переменные. Вполне можно все это заменить на вызов одной функции.

Ovsyanka commented on src/TaskStrategy.php:
```diff
@@ -74,37 +74,49 @@ public function getNextStatus($action)
     }
 
-    public function getAvailableActions(TaskStrategy $obj, $status)
+    public function getAvailableActions($idCurrentUser)
```
Внутри много дублирования кода. Тебе это понадобилось, в частности, из-за того, что разные action принимают разные значения в первом параметре. А вот если все они будут работать одинаково - можно стричь их под одну гребенку =)
Например, можно сделать цикл по всем возможным actions и по каждой проверять доступна ли она для текущего статуса (потребуется доработка класса, например, написание доп. функции для этого) и для текущего пользователя.

Ovsyanka reviewed (changes requested):
В сумме:

Задание практически выполнено. С функциональной точки зрения ошибка только в тестах в функции сравнения массивов. И это прекрасно!

Один из критериев оценки проектов - это использование существующих функций, если они есть. В php как минимум одна подходящая функция для сравнения массивов есть (если не найдешь или будешь сомневаться - спрашивай). Так что нужно использовать ее. При этом твою функцию можно оставить, а внутри нее использовать php функцию. Зачем так делать можно обсудить на консультации.

С точки зрения качества кода есть над чем поработать. Особенно важно исправить то, что явно не соответствет критериям:
* checkRightUser единообразие (честно говоря, конкретный критерий в голову не приходит, но исправить надо)
* дублирование кода в тестах и в getAvailableActions
* использование php функций для сравнения массивов

Все остальное менее важно и больше про то, как можно улучшить качество кода.

Хочу отметить, что тесты проверяют функционал метода getAvailableActions и сейчас будет как раз случай рефакторинга, когда тебе нужно будет изменить код функции (поменяется вызов getRightUser), но при этом функционал ее не меняется. То есть после своих изменений ты просто запустишь тест и он должен точно так же проходить как и раньше.

Было бы неплохо вынести тесты в отдельную директорию и разделить на файлы, в каждом из которых выполняется однотипный набор тестов. В одном проверка на nextStatus, во втором на getAvailableActions. Для таких вещей вообще используется phpunit, но это так, к сведению, в курсе он не проходится.

kind-chigik committed:
Изменены аргументы и видимость методов

- В методе checkRightUser id пользователей теперь передаются по отдельности.

- Метод getInsideName изменен на статичный.

kind-chigik committed:
Сокращен код метода getAvailableActions, удалены лишние if

kind-chigik committed:
Исправлен код теста метода getAvailableActions

- Удалена лишняя переменная idCurrentUser.

- Значения ключей "expected" в тестируемых массивах изменены на внутренние имена классов действий.

- Переписана функция для сравнения получившегося массива действий с ожидаемым.
